File: backend/store/pillar_vector_store.py
"""
PILLAR VECTOR STORE
y = m * x^n
  m = cosine similarity (pillar match strength)
  x = pillar matrix flattened
  n = keyword count

Tracks co-occurrence and correlations between pillars over time.
"""

import math
import time
from typing import List, Dict
from classifier.pillar_classifier import ClassifiedInput

_registry:    Dict[str, dict] = {}
_correlations: Dict[str, dict] = {}


def register_pillars(classified: ClassifiedInput, session_id: str = "", model: str = ""):
    active = [classified.core, classified.emotion, classified.functional] + classified.modifiers
    active = [p for p in active if p]

    for pillar in active:
        if pillar not in _registry:
            _registry[pillar] = {"count": 0, "weight": 1.0,
                                  "last_seen": time.time(),
                                  "sessions": set(), "models": set()}
        _registry[pillar]["count"]    += 1
        _registry[pillar]["last_seen"] = time.time()
        # y = m * x^n → weight grows logarithmically with count
        _registry[pillar]["weight"]    = classified.core_score * math.log1p(_registry[pillar]["count"])
        if session_id:
            _registry[pillar]["sessions"].add(session_id)
        if model:
            _registry[pillar]["models"].add(model)

    # Pairwise correlations
    for i in range(len(active)):
        for j in range(i + 1, len(active)):
            key = "_".join(sorted([active[i], active[j]]))
            if key not in _correlations:
                _correlations[key] = {"count": 0, "strength": 0.0}
            _correlations[key]["count"]    += 1
            _correlations[key]["strength"]  = math.log1p(_correlations[key]["count"])


def get_top_correlations(pillar: str, top_k: int = 3) -> List[dict]:
    results = []
    for key, data in _correlations.items():
        parts = key.split("_")
        if pillar in parts:
            other = "_".join(p for p in parts if p != pillar)
            results.append({"pillar": other, **data})
    return sorted(results, key=lambda x: x["strength"], reverse=True)[:top_k]


def get_dominant_pillars(top_k: int = 5) -> List[dict]:
    return sorted(
        [{"pillar": p, "count": d["count"], "weight": d["weight"]}
         for p, d in _registry.items()],
        key=lambda x: x["weight"], reverse=True
    )[:top_k]


def get_registry()     -> dict: return {k: {**v, "sessions": list(v.get("sessions", set())), "models": list(v.get("models", set()))} for k, v in _registry.items()}
def get_correlations() -> dict: return dict(_correlations)

File: src/store/PillarVectorStore.js
/**
 * PILLAR VECTOR STORE
 * Tracks pillar vectors and their cross-pillar correlations over time.
 *
 * Formula: y = m·x^n
 *   x = pillar matrix
 *   m = slope (intensity/weight of the pillar)
 *   n = count (how many times this pillar has appeared)
 *
 * Cosine similarity tracks which pillars co-occur and correlate.
 */

// Pillar occurrence history: { 'FINANCE': { count, weight, lastSeen } }
let pillarRegistry = {}

// Correlation matrix: { 'FINANCE_STRESS': { count, strength } }
let correlationMatrix = {}

// ── Register a classified input ───────────────────────────────────────────────

export function registerPillars(classified) {
  const { core, emotion, functional, modifiers } = classified
  const activePillars = [core, emotion, functional, ...modifiers].filter(Boolean)

  // Update occurrence counts
  activePillars.forEach(pillar => {
    if (!pillarRegistry[pillar]) {
      pillarRegistry[pillar] = { count: 0, weight: 1.0, lastSeen: Date.now() }
    }
    pillarRegistry[pillar].count    += 1
    pillarRegistry[pillar].lastSeen  = Date.now()
    // weight = m·x^n (slope × count^n, simplified as log growth)
    pillarRegistry[pillar].weight    = Math.log1p(pillarRegistry[pillar].count)
  })

  // Track co-occurrences (pairwise correlations)
  for (let i = 0; i < activePillars.length; i++) {
    for (let j = i + 1; j < activePillars.length; j++) {
      const key = [activePillars[i], activePillars[j]].sort().join('_')
      if (!correlationMatrix[key]) {
        correlationMatrix[key] = { count: 0, strength: 0 }
      }
      correlationMatrix[key].count    += 1
      correlationMatrix[key].strength  = Math.log1p(correlationMatrix[key].count)
    }
  }
}

// ── Query correlations ────────────────────────────────────────────────────────

/**
 * Get the strongest correlated pillar for a given pillar.
 * e.g. "FINANCE" → "STRESS" if they frequently co-occur.
 */
export function getTopCorrelations(pillar, topK = 3) {
  return Object.entries(correlationMatrix)
    .filter(([key]) => key.includes(pillar))
    .sort((a, b) => b[1].strength - a[1].strength)
    .slice(0, topK)
    .map(([key, data]) => {
      const other = key.replace(pillar + '_', '').replace('_' + pillar, '')
      return { pillar: other, ...data }
    })
}

/**
 * Get all pillars ranked by weight (most dominant first).
 */
export function getDominantPillars(topK = 5) {
  return Object.entries(pillarRegistry)
    .sort((a, b) => b[1].weight - a[1].weight)
    .slice(0, topK)
    .map(([pillar, data]) => ({ pillar, ...data }))
}

// ── Cosine similarity between two pillar vectors ──────────────────────────────

/**
 * Compare two pillar sets as binary presence vectors.
 * Used to find how similar two conversations are across pillar space.
 */
export function pillarCosineSimilarity(pillarsA, pillarsB) {
  const allPillars = [...new Set([...pillarsA, ...pillarsB])]
  const vecA = allPillars.map(p => pillarsA.includes(p) ? 1 : 0)
  const vecB = allPillars.map(p => pillarsB.includes(p) ? 1 : 0)

  const dot  = vecA.reduce((sum, v, i) => sum + v * vecB[i], 0)
  const magA = Math.sqrt(vecA.reduce((sum, v) => sum + v ** 2, 0))
  const magB = Math.sqrt(vecB.reduce((sum, v) => sum + v ** 2, 0))

  if (magA === 0 || magB === 0) return 0
  return dot / (magA * magB)
}

// ── Getters ───────────────────────────────────────────────────────────────────

export function getPillarRegistry() { return { ...pillarRegistry } }
export function getCorrelationMatrix() { return { ...correlationMatrix } }

export function resetStore() {
  pillarRegistry    = {}
  correlationMatrix = {}
}
